### PR TITLE
APIView code file token tsp and json schema

### PR DESCRIPTION
**Pull request by praveenkuttappan**
Here is the token schema based on the discussion with @tjprescott and @JonathanGiles. This schema is closer to older token file but at the same time we will have parent child relationship between the code lines to have the context details for APIView. Some other changes are to remove new line and white space tokens as well as to generate navigation TOC by APIView so parser does not haver to generate it.

### CODE REVIEW

**Comment by praveenkuttappan** [comment]
I will submit a separate PR to remove https://github.com/Azure/azure-sdk-tools/tree/main/tools/apiview/parsers/apiview-treestyle-parser-schema once new token model is signed off.

File: tools/apiview/token-schema/model.tsp
Comment (maririos) [inline tools/apiview/token-schema/model.tsp:1]: I liked the location of the previous file. https://github.com/Azure/azure-sdk-tools/tree/main/tools/apiview/parsers 
that way it is next to the other parsers and we could use the [contributing](https://github.com/Azure/azure-sdk-tools/blob/main/tools/apiview/parsers/CONTRIBUTING.md) to explain how to consume it, test, etc

**Comment by maririos** [comment]
Can we generate examples too to see how it will look?

File: tools/apiview/token-schema/model.tsp
Comment (maririos) [inline tools/apiview/token-schema/model.tsp:49]:
```suggestion
   *  are usually not required so they can be excluded when comparing two revisions to avoid reporting them as API changes*/
```

File: tools/apiview/token-schema/model.tsp
Comment (praveenkuttappan) [reply tools/apiview/token-schema/model.tsp:1]: token schema is also consumed by TypeSpec emitter which is aa sibling of `parsers`

**Comment by praveenkuttappan** [comment]
> Can we generate examples too to see how it will look?

As @lmazuel mentioned on another PR, I am trying to use TypeSpec C# emitter to directly generate the .NET models from TypeSpec and then I will create sample token file for Template. We can include this sample file along with the schema for comparison.

File: tools/apiview/token-schema/model.tsp
Comment (tjprescott) [inline tools/apiview/token-schema/model.tsp:62]: Assuming this is the equivalent of `RenderClasses`?

File: tools/apiview/token-schema/model.tsp
Comment (tjprescott) [inline tools/apiview/token-schema/model.tsp:54]: Why is "isHidden" on the token rather than the line? I ask because I'm thinking about how to recreate the TypeSpec navigation. I would have some top-level CodeLines for Resources, Models, etc. but these should not appear in the actual text of the APIView. They _will_ have children. So it seems to me that the CodeLine should be hidden (or not), and if the line is hidden, it will be important to factor that into the automatic indentation. 

Are there use cases for hidden tokens within a line the way there was for hiding things from the diff?

File: tools/apiview/token-schema/model.tsp
Comment (praveenkuttappan) [reply tools/apiview/token-schema/model.tsp]: I will update the docstring and property names as well. diagnostic will be at code file level as it used to and you can link it to line ID.

File: tools/apiview/token-schema/model.tsp
Comment (praveenkuttappan) [reply tools/apiview/token-schema/model.tsp:54]: I don't see any special use case. I can move it to line model.

File: tools/apiview/token-schema/model.tsp
Comment (praveenkuttappan) [reply tools/apiview/token-schema/model.tsp:62]: yes. I will rename it.

File: tools/apiview/token-schema/model.tsp
Comment (maririos) [reply tools/apiview/token-schema/model.tsp:1]: > token schema is also consumed by TypeSpec emitter which is aa sibling of `parsers`

Still don't see this as a problem. the purpose of this file is to be able to communicate the API with the parsers. It would make more sense to have it there too. Also not a hill I will die on :)

Comment (praveenkuttappan) [reply tools/apiview/token-schema/model.tsp:1]: Done.

**Comment by maririos** [comment]
> > Can we generate examples too to see how it will look?
> 
> @maririos @tjprescott I have added a sample token file generated for .NET Azure.Template package and rendered review content text file from this sample token(without docs and system generated comments).

This is great! we should improve then the documentation on https://github.com/Azure/azure-sdk-tools/blob/main/tools/apiview/parsers/CONTRIBUTING.md with this info and how to use the json schema to validate the parser files

**Comment by praveenkuttappan** [comment]
> > > Can we generate examples too to see how it will look?
> > 
> > 
> > @maririos @tjprescott I have added a sample token file generated for .NET Azure.Template package and rendered review content text file from this sample token(without docs and system generated comments).
> 
> This is great! we should improve then the documentation on https://github.com/Azure/azure-sdk-tools/blob/main/tools/apiview/parsers/CONTRIBUTING.md with this info and how to use the json schema to validate the parser files

Updated contributing guide and info about json schema validator.

File: tools/apiview/parsers/CONTRIBUTING.md
Comment (maririos) [inline tools/apiview/parsers/CONTRIBUTING.md:12]: you are removing this because it is not yet implemented or because with the proposed changes it won't be possible

File: tools/apiview/parsers/CONTRIBUTING.md
Comment (maririos) [inline tools/apiview/parsers/CONTRIBUTING.md:16]: maybe?
```suggestion
APIView token schema is available in [TypeSpec](./apiview-treestyle-parser-schema/model.tsp) and [JSON](./apiview-treestyle-parser-schema/model.json). Language parsers needs to create a `CodeFile` object as per the schema.
```

File: tools/apiview/parsers/CONTRIBUTING.md
Comment (maririos) [inline tools/apiview/parsers/CONTRIBUTING.md:17]:
```suggestion
`CodeFile` object contains metadata about the package and an array of `ReviewLine` objects. Each object of `ReviewLine` in `CodeFile` object is a top-level line to be listed. For e.g. Top-level lines are mostly namespace or module level nodes.
```

File: tools/apiview/parsers/CONTRIBUTING.md
Comment (maririos) [inline tools/apiview/parsers/CONTRIBUTING.md:22]:
```suggestion
APIView generates a navigation tree based on the information in the tokens. A token is included in the navigation tree if `NavigatonDisplayName` is set in `ReviewToken` and `LineId` is set in `ReviewLine` object that contains the `ReviewToken`
```

File: tools/apiview/parsers/CONTRIBUTING.md
Comment (maririos) [inline tools/apiview/parsers/CONTRIBUTING.md:28]:
```suggestion
Don't worry about indentation as it will be handled by the tree structure based on the parent-child relationship among `ReviewLine` objects.
```

File: tools/apiview/parsers/CONTRIBUTING.md
Comment (maririos) [inline tools/apiview/parsers/CONTRIBUTING.md:40]:
```suggestion
- Generate APIView token file and paste generated JSON content onto right side panel to validate.
```

File: tools/apiview/parsers/CONTRIBUTING.md
Comment (maririos) [inline tools/apiview/parsers/CONTRIBUTING.md:39]: Jonathan was saying we can publish the json schema on https://www.schemastore.org/json/
or somewhere and that way people can just reference and there is no need for them to manually copy it every time they want to validate it.
Look in Teams for the `APIView Documentation` group chat. Jonathan sent the message on 7/23

File: tools/apiview/parsers/apiview-treestyle-parser-schema/model.tsp
Comment (maririos) [inline tools/apiview/parsers/apiview-treestyle-parser-schema/model.tsp:13]: For my own curiosity and my lack of knowledge here 😨 , we support Java with variant Android, and also Kotlin? do people use both? or how does it work?

File: tools/apiview/parsers/apiview-treestyle-parser-schema/model.tsp
Comment (maririos) [inline tools/apiview/parsers/apiview-treestyle-parser-schema/model.tsp:27]: wording here seems strange. is the `lineID` assigned to review comment and navigation ID? or is it use to compare against those?

File: tools/apiview/parsers/apiview-treestyle-parser-schema/model.tsp
Comment (maririos) [inline tools/apiview/parsers/apiview-treestyle-parser-schema/model.tsp:37]:
```suggestion
  /** Set current line as hidden code line by default. .NET has hidden APIs and architects don't want to see them by default. */
```

File: tools/apiview/parsers/apiview-treestyle-parser-schema/model.tsp
Comment (maririos) [inline tools/apiview/parsers/apiview-treestyle-parser-schema/model.tsp:68]:
```suggestion
  /** Diagnostic ID is auto generated ID by CSharp analyzer. */
```

File: tools/apiview/parsers/apiview-treestyle-parser-schema/model.tsp
Comment (maririos) [inline tools/apiview/parsers/apiview-treestyle-parser-schema/model.tsp:72]:
```suggestion
  /** Auto generated system comment to be displayed under targeted line. */
```

File: tools/apiview/parsers/apiview-treestyle-parser-schema/model.tsp
Comment (maririos) [inline tools/apiview/parsers/apiview-treestyle-parser-schema/model.tsp:70]:
```suggestion
    /** Id of ReviewLine object where this diagnostic needs to be displayed */
```

File: tools/apiview/parsers/apiview-treestyle-parser-schema/model.tsp
Comment (maririos) [inline tools/apiview/parsers/apiview-treestyle-parser-schema/model.tsp:93]: it might be good to add the documentation that says that fatal will block an API from being approved unless strictly overwritten by approves

File: tools/apiview/parsers/CONTRIBUTING.md
Comment (praveenkuttappan) [reply tools/apiview/parsers/CONTRIBUTING.md:12]: > you are removing this because it is not yet implemented or because with the proposed changes it won't be possible

It's not added yet. I am removing it since it's not a benefit yet straight from this new token model. We can add it once the feature is added in APIView.

File: tools/apiview/parsers/apiview-treestyle-parser-schema/model.tsp
Comment (praveenkuttappan) [reply tools/apiview/parsers/apiview-treestyle-parser-schema/model.tsp:13]: It's handled by Java parser.

File: tools/apiview/parsers/apiview-treestyle-parser-schema/model.tsp
Comment (praveenkuttappan) [reply tools/apiview/parsers/apiview-treestyle-parser-schema/model.tsp:27]: Line ID is not used for comparison. It's only used to identify a line from a given Line ID so comments use it to identify where a comment needs to be listed. Comment stores this Line ID in DB. Similarly, Line ID is used as HREF within the review so a type name can navigate to its type definition.

**Review by maririos — APPROVED** [review]

File: tools/apiview/parsers/CONTRIBUTING.md
Comment (praveenkuttappan) [reply tools/apiview/parsers/CONTRIBUTING.md:39]: I have discussed this with Jonathan and he is fine with JSON schema stored in azure-sdk-tools repo.

**Pull request merged**
